src/dvc_data/fs.py

```diff
@@ -71,23 +71,38 @@ def open(  # type: ignore
         return fs.open(fspath, mode=mode, encoding=encoding)
 
     def ls(self, path, detail=True, **kwargs):
-        info = self.info(path)
-        if info["type"] != "directory":
-            return [info] if detail else [path]
-
         root_key = self._get_key(path)
-        try:
-            entries = [
-                self.path.join(path, name) if path else name
-                for name in self.index.ls(prefix=root_key)
-            ]
-        except KeyError as exc:
-            raise FileNotFoundError from exc
+        info = self.index.info(root_key)
+        if info["type"] != "directory":
+            if detail:
+                info["name"] = path
+                return [info]
+            else:
+                return [path]
 
+        self.index._ensure_loaded(root_key)  # pylint: disable=protected-access
         if not detail:
-            return entries
 
-        return [self.info(epath) for epath in entries]
+            def node_factory(_, key, children, *args):
+                if key == root_key:
+                    list(children)
+                else:
+                    entries.append(self.path.join(path, key[-1]))
+
+        else:
+
+            def node_factory(_, key, children, *args):
+                if key == root_key:
+                    list(children)
+                else:
+                    epath = self.path.join(path, key[-1])
+                    info = self.index.info(key)
+                    info["name"] = epath
+                    entries.append(info)
+
+        entries = []
+        self.index.traverse(node_factory, prefix=root_key)
+        return entries
```

Comment:
Would this fit into `index.ls(detail=...)`?

Reply:
Yes, I agree, except that `.ls()` needs to compute the path for the entry, which the index can't do.

Reply:
@rlamy Ah, right. But in index sense the path is the key, so maybe it could return that and we would then process it in fs.

Reply:
In that case, `index.ls()` should probably return an iterator.

Reply:
Sounds great to me 🙂